[DOCS] Add "Language Influences" page (#11198)
ethereum/solidity: docs/language-influences.rst

###################
Language Influences
###################

> The Solidity language is a `curly-bracket language <https://en.wikipedia.org/wiki/List_of_programming_languages_by_type#Curly-bracket_languages>`_

Reviewer: Not sure if this is better:
```suggestion
Solidity is a `curly-bracket language <https://en.wikipedia.org/wiki/List_of_programming_languages_by_type#Curly-bracket_languages>`_
that has been influenced and inspired by several well-known programming languages.
```

Author: yes :)

and has been influenced and inspired by several well-known programming languages.

> Solidity is most profoundly influenced by C++, but also borrowed concepts from other languages like

Reviewer: "other" and "others" in the same sentence:
```suggestion
Solidity is most profoundly influenced by C++, but also borrowed concepts from languages like
```

Python, JavaScript, and others.

> 

Reviewer 2: Maybe we should go a bit into detail about the influence from C++? Although I don't really know what to say there.

Reviewer 2: ``The influence from C++ can be seen in the syntax for variable declarations, for loops, the concept of overloading functions, implicit and explicit type conversions and many other details.``

In the early days of the language, Solidity used to be partly influenced by JavaScript.
This was due to function-level scoping of variables and the use of the keyword ``var``.

> The JavaScript influence was more and more reduced starting from version 0.4.0.

Reviewer 3: ```suggestion
The JavaScript influence was reduced starting from version 0.4.0.
```

Now, the main similarity to JavaScript is that functions are defined using the keyword
``function``. Solidity also supports import syntax and semantics that
are similar to those available in JavaScript. Besides those points, Solidity looks like

> most other curly-bracket languages and has no major JavaScript influence anymore.

Reviewer 3: This sentence comes as we are desperately trying to avoid any association with javascript :)

Author: But I think we still have to close the narrative somehow since we start the section with "it used to be influenced" etc.

Another influence to Solidity was Python. Solidity's modifiers were added trying to model

> Python's decorators with a much more restricted functionality. Furthermore, multiple inheritance

Reviewer 3: ```suggestion
Python's decorators with a much more restricted functionality. Furthermore, multiple inheritance, C3 linearization,
```

> and the ``super`` keyword are taken from Python as well as the general value semantics of value

Reviewer 3: ```suggestion
and the ``super`` keyword are taken from Python as well as the general semantics of value
```

> and the ``super`` keyword are taken from Python as well as the general value semantics of value

Reviewer: super is also present in Javascript.

Author: What should I do with this information? Where was it first?

and reference types.

> 

Reviewer 3: We could add that several Yul optimizer steps have been inspired by llvm's optimization steps. Not exactly about language though. Many of them are generic, and can't really be attributed to llvm. So I'm not sure about this any more.

Reviewer 2: This could rather fit into the optimizer section, I would say.

> Last but not least, ``payable`` was inspired by an, we believe, `unnamed language <https://publications.lib.chalmers.se/records/fulltext/234939/234939.pdf>`_

Reviewer 3: ```suggestion
Last but not least, ``payable`` was inspired by, we believe, `unnamed language <https://publications.lib.chalmers.se/records/fulltext/234939/234939.pdf>`_
```

> Last but not least, ``payable`` was inspired by an, we believe, `unnamed language <https://publications.lib.chalmers.se/records/fulltext/234939/234939.pdf>`_

Reviewer: Not sure this is really relevant.

Author: I just took what was there in the PR and Chris had listed it. :)

Reviewer: Yeah but we debated there too :)

Author: Ok so @chriseth what do you think? Keep or delete?

Author: Ok so @chriseth wdyt? Keep it or delete it?

Reviewer 2: I'm fine with either.

###################
Language Influences
###################

Solidity is a `curly-bracket language <https://en.wikipedia.org/wiki/List_of_programming_languages_by_type#Curly-bracket_languages>`_
that has been influenced and inspired by several well-known programming languages.

> and has been influenced and inspired by several well-known programming languages.

Reviewer 2: duplicate line

Solidity is most profoundly influenced by C++, but also borrowed concepts from languages like
Python, JavaScript, and others.

In the early days of the language, Solidity used to be partly influenced by JavaScript.
This was due to function-level scoping of variables and the use of the keyword ``var``.
The JavaScript influence was reduced starting from version 0.4.0.

> Now, the main similarity to JavaScript is that functions are defined using the keyword

Reviewer 2: Maybe "the main remaining similarity to JavaScript"?

``function``. Solidity also supports import syntax and semantics that
are similar to those available in JavaScript. Besides those points, Solidity looks like
most other curly-bracket languages and has no major JavaScript influence anymore.

Another influence to Solidity was Python. Solidity's modifiers were added trying to model
Python's decorators with a much more restricted functionality. Furthermore, multiple inheritance, C3 linearization,

> and the ``super`` keyword are taken from Python as well as the general semantics of value

Reviewer 2: ```suggestion
and the ``super`` keyword are taken from Python as well as the general assignment and copy semantics of value
```
